src/saving/ISaveUpgradeStep.cs

```diff
@@ -162,6 +163,174 @@ private void UpgradeBehaviouralValues(JProperty property, JEnumerable<JProperty>
+
+                    if (adjacentChildrenNames.Contains("Depth") && !adjacentChildrenNames.Contains("Volume"))
+                    {
+                        UpgradePatchesVolume((JObject)adjacent, adjacentChildren);
+                    }
+                }
+            }
+
+            // Add gas production effect in effects array
+            if (property.Name == "effects")
+            {
+                var gasProductionEffectName = "GasProductionEffect, Thrive";
+                if (!PropertyHasNamedEffect(property, gasProductionEffectName))
+                {
+                    UpgradeGasProductionEffect(property, gasProductionEffectName);
+                }
+            }
+        }
+
+        private bool PropertyHasNamedEffect(JProperty property, string name)
+        {
+            if (property.Value is JArray tokenArray)
+            {
+                foreach (var effect in tokenArray)
+                {
+                    var effectProperties = effect.Children<JProperty>();
+
+                    if (effectProperties.Select(p => p.Value.ToString()).Contains(name))
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        ///   Updates patches values by adding volume to it as a cube.
+        /// </summary>
+        /// <param name="container">The JObject containing the patch definition.</param>
+        /// <param name="children">The children of this JObject.</param>
+        /// <remarks>
+        ///   <para>
+        ///     Changes a json either like:
+        ///       "0": {
+        ///         ...
+        ///         "Depth": [2500, 3000],
+        ///         ...
+        ///       }
+        ///     or
+        ///       "Adjacent": [
+        ///         {
+        ///           ...
+        ///           "Depth": [2500, 3000],
+        ///           ...
+        ///         },
+        ///         ...
+        ///       ]
+        ///     to
+        ///       ...
+        ///       {
+        ///         ...
+        ///         "Depth": [2500, 3000],
+        ///         "Volume": (3000 - 2500)^3,
+        ///         ...
+        ///       }
+        ///       ...
+        ///   </para>
+        /// </remarks>
+        private void UpgradePatchesVolume(JObject container, JEnumerable<JProperty> children)
+        {
+            if (children.First(p => p.Name == "Depth").Value is JArray depth)
+            {
+                if (depth[1].Value<int?>() == null || depth[0].Value<int?>() == null)
+                {
+                    throw new ArgumentException("Depth array does not start with two integers!");
+                }
+
+                var depthDifference = depth[1].Value<int>() - depth[0].Value<int>();
+
+                // Assume cubic patches for upgrade
+                container.Add("Volume", (long)Math.Pow(depthDifference, 3));
```

Comment:
Why is this done like this? Surely it would be better to either: specify the volume in the biome definitions *or* also update the patch generator to use the same depth difference logic, otherwise I fear there's too much highly specific math duplication between this and the patch generation.

Reply:
Patch generator is hardcoded. I used cubes as a simple model for patches, but nothing should force patches to be cubes. Now this upgrade step just intends to give a volume to patches of older versions, and since I can't access the exact value (unless I just move every patch volume to `Constants.cs` but then any modification would require to change multiple files at once, which seems terrible an idea, if I may), I just go for the simple model of a cube, which is enough to get older versions to be *playable*.

The only way I see to change this would be to rework the patch generator to work from either Constants or JSONs, and that would be a PR of its own.

Reply:
I more suggested that you add the "Volume" as a parameter for a biome, then use that as the value here and in the patch map generator. I guess if we don't care about the math being off (and thus perhaps balancing as well) when comparing upgraded and other saves, this is fine if that is explained with a comment why this approach is taken.

Reply:
You mean having a predetermined volume for a biome? I see no reason why we should have a predefined volume for caves : I think this goes along the same logic as the one used for depth measures.  I'll go along the comment route, I think.

Ultimately, I guess this shows how much the current patch generator system should be updated...

Reply:
I was suggesting that as it would fit the current patch generation model and also make it consistent when upgrading saves.
But feel free to do this:
> I guess if we don't care about the math being off (and thus perhaps balancing as well) when comparing upgraded and other saves, this is fine if that is explained with a comment why this approach is taken.